packages/trace-viewer/src/sw/snapshotServer.ts

```diff
@@ -18,24 +18,46 @@ import type { URLSearchParams } from 'url';
 import type { SnapshotRenderer } from './snapshotRenderer';
 import type { SnapshotStorage } from './snapshotStorage';
 import type { ResourceSnapshot } from '@trace/snapshot';
+import type { ContextEntry, PageEntry } from '../types/entries';
+
+function findClosest<T>(items: T[], metric: (v: T) => number, target: number) {
+  return items.find((item, index) => {
+    if (index === items.length - 1)
+      return true;
+    const next = items[index + 1];
+    return Math.abs(metric(item) - target) < Math.abs(metric(next) - target);
+  });
+}
 
 type Point = { x: number, y: number };
 
 export class SnapshotServer {
   private _snapshotStorage: SnapshotStorage;
   private _resourceLoader: (sha1: string) => Promise<Blob | undefined>;
   private _snapshotIds = new Map<string, SnapshotRenderer>();
+  private _pages: Map<string, PageEntry>;
 
-  constructor(snapshotStorage: SnapshotStorage, resourceLoader: (sha1: string) => Promise<Blob | undefined>) {
+  constructor(snapshotStorage: SnapshotStorage, resourceLoader: (sha1: string) => Promise<Blob | undefined>, contextEntries: ContextEntry[]) {
     this._snapshotStorage = snapshotStorage;
     this._resourceLoader = resourceLoader;
+    this._pages = new Map(contextEntries.flatMap(c => c.pages.map(p => [p.pageId, p])));
   }
 
-  serveSnapshot(pathname: string, searchParams: URLSearchParams, snapshotUrl: string): Response {
+  serveSnapshot(pathname: string, searchParams: URLSearchParams, snapshotUrl: string, swScope: string): Response {
     const snapshot = this._snapshot(pathname.substring('/snapshot'.length), searchParams);
     if (!snapshot)
       return new Response(null, { status: 404 });
-    const renderedSnapshot = snapshot.render();
+
+    let screenshotUrl: URL | undefined;
+    const { wallTime, timestamp, pageId } = snapshot.snapshot();
+    const page = this._pages.get(pageId);
+    if (page) {
+      const closestFrame = (wallTime && page.screencastFrames[0]?.frameSwapWallTime) ? findClosest(page.screencastFrames, frame => frame.frameSwapWallTime!, wallTime) : findClosest(page.screencastFrames, frame => frame.timestamp, timestamp);
```

Comment:
I'd prefer this code to live in `SnapshotRenderer` instead.

Reply:
Would `SnapshotRenderer` get access to `screencastFrames`? Would we pass them into `.render()`, or do we pass them into the constructor?

Reply:
I've since moved the code a little, but it's still inside the server. Let me know how to proceed.

Reply:
we discussed this offline and ended up with https://github.com/microsoft/playwright/pull/33119/commits/45d1b54226e31615e26bbdb6e47830c4cc665e48.